### PR TITLE
WIP: treewide: Always prefix compilers

### LINKED ISSUE #21471: Always cross compile
Exherbo, another linux distribution that has done good work with cross-compiling, always builds cross compilers---i.e. `--host` `--build` and `--target` are always passed to gcc even if some of those flags are the same: http://git.exherbo.org/arbor.git/tree/packages/sys-devel/gcc/gcc.exlib#n187. The big advantage here is by using the cross-compiling code path in all cases, there's less to maintain, and actual cross-compiling is less likely to rot.

In https://github.com/NixOS/nixpkgs/pull/21268 (specifically https://github.com/NixOS/nixpkgs/pull/21268/commits/633feb4e39a24c7da9ff4ae1b5f4219e0c2a1bb6 but I'll probably rebase at some point breaking that link), I introduce always-defined `buildPlatform` `hostPlatform` and `targetPlatform` for the same reasons (I keep around nullable `crossSystem` and not-always present `stdenv.cross` for compatibility). So if/when that PR is merged would be a good time to tackle this.

Naturally this would be a gcc-caused mass rebuild on Linux. Darwin shouldn't be as affected as LLVM, by default, includes all the targets we care about.

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

The goal is to bring the native and cross code paths closer together by always prefixing compilers. This makes it easier to catch programs that hard-code `gcc` or `clang`.

This also includes:

  - Two patches to help with darwin -> linux cross compilation; since they're fine patches and the same override-`gcc` fix is needed.
  - Making the wrapper env vars define `CC`, `LD`, and friends use full paths instead of names, generally good for robustness.

The whole thing I plan to merge after 18.09, but maybe those 2 bits can go in first.

Fixes #21471

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/paxctl/default.nix [hunk 1]
```diff
@@ -10,7 +10,9 @@ stdenv.mkDerivation rec {
   };
 
   preBuild = ''
-    sed "s|--owner 0 --group 0||g" -i Makefile
+    sed -i Makefile \
+      -e 's|--owner 0 --group 0||g' \
+      -e '/CC:=gcc/d'
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply pkgs/os-specific/linux/paxctl/default.nix]: No I don't. But it's part of bootstrapping; that's why I fixed it.

**Comment by Ericson2314** [comment]
@LnL7 actually wrapped binaries are always prefixed with this PR. But I could make `targetPrefix` a parameter. `inPrefix` is added because GCC will be wrapped on the inside but Clang won't be, but yes they can be varied independently in principle.

**Comment by dezgeg** [comment]
NAK for removing unprefixed `gcc`, `clang` from native compilation. `stdenv` stands for "standard environment", and that's what it should stay, standard.

File: pkgs/os-specific/linux/kernel-headers/default.nix [hunk 2]
```diff
@@ -23,16 +23,35 @@ let
...
+      "cc-version:=9999"
+      "cc-fullversion:=999999"
+      # `$(..)` expanded by make alone
+      "HOSTCC:=$(BUILD_CC)"
+      "HOSTCXX:=$(BUILD_CXX)"
+    ];
+
+    # Skip clean on darwin, case-sensitivity issues.
+    buildPhase = lib.optionalString (!buildPlatform.isDarwin) ''
+      make mrproper $makeFlags
```
Comment (matthewbauer) [inline pkgs/os-specific/linux/kernel-headers/default.nix]: Can you use buildFlags & checkFlags here instead?

**Comment by matthewbauer** [comment]
Can we have an unprefixed symlink for the host compiler? Even just something like:

```
ln -s $CC $out/bin/cc
ln -s $CXX $out/bin/c++
```

Would cover most use cases. Maybe one for gcc as well because that's often hardcoded.

**Comment by Ekleog** [comment]
I completely agree with the gain offered by always prefixing the compilers in stdenv: it is the removal of another source of impurity (the host system), which is what nix is all about.

However, a cursory read (which can easily be wrong) makes me think `nix-shell -p gcc` would also only yield prefixed-gcc, which would not be acceptable as `gcc` is then supposed to be used by the end-user, and not by nix. Did I read correctly?

**Comment by Ericson2314** [comment]
`gcc` and `stdenv.cc` are currently the same thing (for Linux). I didn't change that but we could.

I sort of rather fix that after as `gcc` has a bunch of other issues for manual usage too.

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Comment by cdepillabout** [comment]
It looks like there hasn't been any updates on this in a year, so I will close it.

(@Ericson2314 sorry in advance if you're still working on any of these PRs that I've closed.  Definitely don't hesitate to re-open.  I've been trying to clean up old PRs tagged `haskell`.)

**Comment by Ericson2314** [comment]
@cdepillabout It's fine. I'll reopen if/when I get back to them.

**Pull request closed without merging**